Run each subscription's packet processing in its own worker

daniel-abramov opened the pull request:
I've realised that actually that generic helper that I created for the subscriptions to track missing packets (watchdog) is very close in semantics to the concept of a single worker. In fact, the watchdog is just a special case of a worker where the task is an empty `{}` and the task handler is an empty function. So I decided to generalize it a bit and that almost automatically allowed me to transform watchdog into a subscription worker which handles incoming packets. So it was a fairly elegant transformation that did not touch many parts of the code 🙂 

So now, each call to the `Subscription.WriteRTP()` does not block, but merely sends the packet to the worker that then handles the packets.

I've also made it non-blocking since if the channel blocks in this case, it means that the worker can't keep up with the load and so packet drops are inevitable anyway.

Fixes https://github.com/matrix-org/waterfall/issues/108.

daniel-abramov committed:
worker: rename `watchdog` -> `worker`

We'll need it to generalise the watchdog (worker) logic.

daniel-abramov committed:
worker: generalise the worker to accept tasks

daniel-abramov committed:
worker: make sends non-blocking

daniel-abramov committed:
worker: make channel size configurable

daniel-abramov committed:
subscription(video): handle packets in worker

Fixes https://github.com/matrix-org/waterfall/issues/108.

dbkr commented on pkg/common/worker.go:
```diff
@@ -0,0 +1,91 @@
+		case c.channel <- task:
+			return nil
+		default:
+			return ErrWorkerTooBusy
```
Will this risk dropping the most recent packets and therefore the SFU ending up permanently behind? We presumably still need to keep this a last resort, but probably need the worker OnTask function to drop the oldest packets in the queue in order to catch up?

daniel-abramov replied:
Yes, you're right, the approach that you suggest is also something that I think is a better choice. The only reason I did not implement it is that I thought that it might be a bit complicated (especially with packet rewriting), so it probably requires a bit more thinking, so I thought a "smarter" way (the way you suggested) to drop older packets is better to implement once we land the support of keyframe detection and switching on key frames since it would allow us to implement the logic (drop all non-key-frame packets from the queue when the queue gets full too quickly).

One thing to note is that if we really return `ErrWorkerTooBusy`, then it would mean that the conference is quite overloaded already and the SFU can't keep up with the new subscriptions. So ideally it should not happen at all for conferences, I may assume that we may get this error only when there are too many things going on in the SFU.

daniel-abramov committed:
worker: don't use word "watchdog" anymore

So that it's not confusing (we renamed the watchog component to a
worker, so now it makes sense to use a new terminology to not cause any
confusion).

EnricoSchw commented on pkg/common/worker_test.go:
```diff
@@ -0,0 +1,23 @@
+	w := common.StartWorker(workerConfig)
+
+	for n := 0; n < b.N; n++ {
+		w.Send(struct{}{})
```
This always returns an error, because we have to add an buffer for the channel in the WorkerConfig.

daniel-abramov replied:
Oh, good catch! 👍 Thanks!

daniel-abramov committed:
worker(test): fix missing queue size
